File: app/src/main/java/com/aol/philipphofer/logic/sudoku/Sudoku.java
package com.aol.philipphofer.logic.sudoku;

import com.aol.philipphofer.logic.Position;

public class Sudoku {

    private final Block[] blocks;
    public int overallErrors = 0;

    public Sudoku() {
        blocks = new Block[9];

        for (int i = 0; i < 9; i++)
            blocks[i] = new Block();
    }

    public void insert(int number, Position position, boolean note) {
        if (blocks[position.block].insert(number, position, note))
            overallErrors++;
    }

    public void delete(Position position) {
        blocks[position.block].delete(position);
    }

    public int freeFields() {
        int freeFields = 0;
        for (int k = 0; k < 9; k++)
            for (int i = 0; i < 3; i++)
                for (int j = 0; j < 3; j++)
                    if (getNumber(new Position(i, j, k)).getNumber() == 0)
                        freeFields++;
        return freeFields;
    }

    public int currentErrors() {
        int currentErrors = 0;
        for (int k = 0; k < 9; k++)
            for (int i = 0; i < 3; i++)
                for (int j = 0; j < 3; j++)
                    if (getNumber(new Position(i, j, k)).isError())
                        currentErrors++;
        return currentErrors;
    }

    public Number getNumber(Position position) {
        return blocks[position.block].getNumber(position.row, position.column);
    }

    public void setNumber(Position position, Number number) {
        blocks[position.block].setNumber(position.row, position.column, number);
    }

    public Block[] getSudoku() {
        return blocks;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 9; i = i + 3) {
            for (int k = 0; k < 3; k++) {
                for (int j = i; j < i + 3; j++) {
                    for (int a = 0; a < 3; a++)
                        sb.append(blocks[j].getNumbers()[k][a].getNumber());
                    sb.append(" ");
                }
                sb.append("\n");
            }
            sb.append("\n");
        }
        return sb.toString();
    }
}

File: app/src/main/java/com/aol/philipphofer/logic/sudoku/Sudoku.kt
package com.aol.philipphofer.logic.sudoku

import com.aol.philipphofer.logic.Position

class Sudoku {
    val blocks: Array<Block> = Array(9) { Block() }
    var overallErrors = 0

    fun insert(number: Int, position: Position, note: Boolean) {
        if (!blocks[position.block].insert(number, position, note)) overallErrors++
    }

    fun delete(position: Position) {
        blocks[position.block].delete(position)
    }

    fun freeFields(): Int {
        var freeFields = 0
        for (block in 0..8)
            for (row in 0..2)
                for (col in 0..2)
                    if (getNumber(Position(block, row, col)).number == 0)
                        freeFields++

        return freeFields
    }

    fun currentErrors(): Int {
        var currentErrors = 0
        for (k in 0..8)
            for (i in 0..2)
                for (j in 0..2)
                    if (getNumber(Position(k, i, j)).isError)
                        currentErrors++
        return currentErrors
    }

    fun getNumber(position: Position): Number {
        return blocks[position.block].getNumber(position.row, position.column)
    }

    fun setNumber(position: Position, number: Number?) {
        blocks[position.block].setNumber(position.row, position.column, number!!)
    }

    override fun toString(): String {
        val sb = StringBuilder("Sudoku with $overallErrors overall-errors: \n")
        var i = 0
        while (i < 9) {
            for (k in 0..2) {
                for (j in i until i + 3) {
                    for (a in 0..2)
                        sb.append(blocks[j].numbers[k][a].number)
                    sb.append(" ")
                }
                sb.append("\n")
            }
            sb.append("\n")
            i += 3
        }
        return sb.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Sudoku

        if (!blocks.contentDeepEquals(other.blocks)) return false
        if (overallErrors != other.overallErrors) return false

        return true
    }
}